test/csuite/wt7989_compact_checkpoint/main.c

```diff
@@ -0,0 +1,311 @@
+    testutil_check(session->create(session, uri, table_config));
+
+    populate(session, uri);
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range to let compact relocate blocks from the
+     * end of the file.
+     */
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    td.conn = conn;
+    td.uri = uri;
+    td.cond = NULL;
+
+    /* Spawn checkpoint and compact threads. Order is important! */
+    if (stress_test) {
+        testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+        testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+    } else {
+        /* Create and initialize conditional variable. */
+        testutil_check(__wt_cond_alloc((WT_SESSION_IMPL *)session, "compact operation", &td.cond));
+
+        testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+        testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+    }
+
+    /* Wait for the threads to finish the work. */
+    (void)pthread_join(thread_checkpoint, NULL);
+    (void)pthread_join(thread_compact, NULL);
+
+    file_sz_after = get_file_size(session, uri);
+
+    /* Cleanup */
+    if (!stress_test) {
+        __wt_cond_destroy((WT_SESSION_IMPL *)session, &td.cond);
+        td.cond = NULL;
+    }
+
+    testutil_check(session->close(session, NULL));
+
+    /* Check if there's at least 10% compaction. */
+    printf(" - Compressed file size MB: %f\n - Original file size MB: %f\n",
+      file_sz_after / (1024.0 * 1024), file_sz_before / (1024.0 * 1024));
+
+    /*
+     * FIXME-WT-8055 At the moment the assert below is commented out to prevent evergreen from going
+     * red. Please enable the assert as soon as the underlying defect is fixed and compact does its
+     * job well.
+     */
+    /*testutil_assert(file_sz_before * 0.9 > file_sz_after);*/
+}
+
+static void *
+thread_func_compact(void *arg)
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. Sleep for one
+         * second.
+         */
+        __wt_sleep(1, 0);
+
+        /* Wake up the checkpoint thread. */
+        printf("Sending the signal!\n");
+        __wt_cond_signal((WT_SESSION_IMPL *)session, td->cond);
+    }
+
+    /* Perform compact operation. */
+    testutil_check(session->compact(session, td->uri, NULL));
+    testutil_check(session->close(session, NULL));
+
+    return (NULL);
+}
+
+static bool
+wait_run_check(WT_SESSION_IMPL *session)
+{
+    (void)session; /* Unused */
+
+    /* Always return true to make sure __wt_cond_wait_signal does wait. This callback is required
```

Comment:
Multi-line comment not formatted correctly.

Reply:
Done